packages/eslint-plugin-query/src/rules/exhaustive-deps.rule.ts

```diff
@@ -0,0 +1,173 @@
+import type { JSONSchema, TSESLint, TSESTree } from '@typescript-eslint/utils'
+import { AST_NODE_TYPES, ESLintUtils } from '@typescript-eslint/utils'
+import { ExtraUtils } from '../extra-utils'
+
+const messages = {
+  missingDeps:
+    'The following dependencies are missing in your queryKey: {{deps}}',
+}
+
+const ruleSchema: JSONSchema.JSONSchema4 = {
+  type: 'array',
+  minItems: 0,
+  maxItems: 1,
+  items: {
+    type: 'object',
+    properties: {
+      whitelist: {
+        type: 'array',
+        items: {
+          type: 'string',
+        },
+      },
+    },
+  },
+}
+
+const defaultRuleOptions = { whitelist: ['api', 'fetch', 'axios'] }
+
+type RuleMessage = keyof typeof messages
+type RuleOptions = { whitelist?: string[] }
+type RuleContext = Readonly<TSESLint.RuleContext<RuleMessage, [RuleOptions]>>
+
+const createRule = ESLintUtils.RuleCreator(
+  () => 'https://github.com/tanstack/query',
+)
+
+export const exhaustiveDepsRule = createRule<[RuleOptions], RuleMessage>({
+  name: 'exhausitive-deps',
+  meta: {
+    docs: {
+      description:
+        'This rule ensures that all dependencies are passed to the useQuery queryKey parameter',
+      recommended: 'warn',
+    },
+    messages: messages,
+    type: 'problem',
+    fixable: 'code',
+    schema: ruleSchema,
+  },
+  defaultOptions: [defaultRuleOptions],
+  create(context) {
+    return {
+      CallExpression(node) {
+        if (
+          ExtraUtils.isIdentifier(node.callee) &&
+          node.callee.name === 'useQuery'
+        ) {
+          runCheck({ node, context })
+        }
+      },
+    }
+  },
+})
+
+function runCheck(params: {
+  node: TSESTree.CallExpression
+  context: RuleContext
+}) {
+  const { node, context } = params
+  const queryOptions = node.arguments[0]
+  const ruleOptions = { ...defaultRuleOptions, ...context.options[0] }
+
+  if (queryOptions?.type !== AST_NODE_TYPES.ObjectExpression) {
+    return
+  }
+
+  const scopeManager = context.getSourceCode().scopeManager
+  const queryKey = ExtraUtils.findPropertyWithIdentifierKey(
+    queryOptions.properties,
+    'queryKey',
+  )
+  const queryFn = ExtraUtils.findPropertyWithIdentifierKey(
+    queryOptions.properties,
+    'queryFn',
+  )
+
+  if (
+    scopeManager === null ||
+    queryKey === undefined ||
+    queryFn === undefined ||
+    queryFn.value.type !== AST_NODE_TYPES.ArrowFunctionExpression
+  ) {
+    return
+  }
+
+  if (queryKey.value.type !== AST_NODE_TYPES.ArrayExpression) {
+    // TODO support query key factory
+    return
+  }
```

Comment:
wouldn't a rather simple approach at tackling this be to look if the queryKey is a function, and if it is, look at the params of that function?

So:

`useQuery({ queryKey: ["todos", id] })`

would be the same as:

`useQuery({ queryKey: todoKeys.detail(id) })`

they both reference a variable `id`.

Reply:
Correct. But it can be a bit more complex:

```
queryKey: todoKeys.a({ x })
queryKey: todoKeys.a([b, { c }])
```

I need to recursively iterate over the nodes and then determine:

1. If all of the dependencies are there (trivial)
2. If there's a missing dependency, how should I auto-fix it? the function can accept anything. I can idiomatically put it inside the last function expression.

Reply:
I don't think auto-fixing is required if it gets too complex. The order of arrays is important (for example, for invalidation), so you'd likely need to change that anyways (unlike the react hooks rule, where the order doesn't matter).

Reply:
Seems right. I'll write an implementation without an auto fix